### PR TITLE
Speedup compilation times

### PR DESCRIPTION
**Pull request by bjornpiltz**
This PR tries to move as much code as possible away from header files. It specifically removes all references to cereal from public includes. Here are some results:
- Travis build time for all tests, samples and exes has gone from [1h44](https://travis-ci.org/bjornpiltz/openMVG/builds/228716984) to [53m](https://travis-ci.org/bjornpiltz/openMVG/builds/229157648).
- The real elapsed time went from 46 min to 20 min

More importantly the usability of openMVG has gone up. Before, a file containing the line `#include <openMVG/sfm/sfm.hpp>` took 29 s to compile. This is now down to 2.5 seconds.

Some of the main_*.cpp compiles significantly faster. Here are some examples:
image

### CODE REVIEW

**Comment by zwn** [comment]
@bjornpiltz This is a big deal for me. Thanks!

Does it help in any way with #866? I see that you are making changes related to cereal and registration. Does it also remove the exp and lib files by any chance?

**Comment by bjornpiltz** [comment]
@zwn  I haven't checked for size yet. I just realized my build directory is 40 GB! I'm going to have to make a clean build to check that, but right now I only have 40 MB of exp files, so that's not a problem.

File: src/openMVG/sfm/pipelines/structure_from_known_poses/structure_estimator.cpp [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include "openMVG/graph/graph.hpp"
 #include "openMVG/geometry/pose3.hpp"
 //#include "openMVG/matching/defines.hpp"
```
Comment (pmoulon) [inline src/openMVG/sfm/pipelines/structure_from_known_poses/structure_estimator.cpp]: You can remove the unused commented include

File: src/openMVG/sfm/sfm_data_BA_ceres.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 
 #include "ceres/problem.h"
 #include "ceres/solver.h"
+#include "openMVG/numeric/numeric.h"
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_data_BA_ceres.cpp]: include order (alphabetical ordering)

Do we need numeric or just "openMVG/numeric/eigen_alias_definition.hpp"

File: src/openMVG/sfm/sfm_data_filters.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@
 #include "openMVG/tracks/union_find.hpp"
 
 #include <utility>
+#include <iterator>
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_data_filters.cpp]: put iterator before utility

File: src/openMVG/sfm/sfm_data_io_cereal.cpp [hunk 4]
```diff
@@ -23,6 +24,7 @@
 #include <cereal/types/string.hpp>
 #include <cereal/types/unordered_map.hpp>
 #include <cereal/types/vector.hpp>
+#include <cereal/types/polymorphic.hpp>
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_data_io_cereal.cpp]: alphabetical ordering

File: src/openMVG/sfm/sfm_landmark.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+// Copyright (c) 2015 Pierre Moulon.
+
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "openMVG/sfm/sfm_landmark.hpp"
+
+#include <cereal/types/vector.hpp>
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_landmark.cpp]: alphabetical ordering

File: src/openMVG/sfm/sfm_landmark.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+template <class Archive>
+void Observation::load( Archive & ar)
+{
+  ar(cereal::make_nvp("id_feat", id_feat ));
+  std::vector<double> p(2);
+  ar(cereal::make_nvp("x", p));
+  x = Eigen::Map<const Vec2>(&p[0]);
+}
+
+template void Observation::load
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_landmark.cpp]: I wold like to suggest to use alphabetical ordering based on the cereal::X

File: src/openMVG/sfm/sfm_landmark.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+<cereal::PortableBinaryOutputArchive>
+(cereal::PortableBinaryOutputArchive & ar) const;
+
+// Serialization
+template <class Archive>
+void Landmark::save( Archive & ar) const
+{
+  const std::vector<double> point = { X(0), X(1), X(2) };
+  ar(cereal::make_nvp("X", point ));
+  //ar(cereal::make_nvp("observations", obs));
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_landmark.cpp]: You must add back this 
see
https://github.com/openMVG/openMVG/pull/903/files#diff-d046a42108ff624358eaa4402d09c595L64
and https://github.com/openMVG/openMVG/blob/master/src/openMVG/sfm/sfm_landmark.hpp#L60

File: src/openMVG/sfm/sfm_landmark.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+template <class Archive>
+void Landmark::load( Archive & ar)
+{
+  std::vector<double> point(3);
+  ar(cereal::make_nvp("X", point ));
+  X = Eigen::Map<const Vec3>(&point[0]);
+  //ar(cereal::make_nvp("observations", obs));
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_landmark.cpp]: You must add back this https://github.com/openMVG/openMVG/blob/master/src/openMVG/sfm/sfm_landmark.hpp#L69

File: src/openMVG/sfm/sfm_view.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+// Copyright (c) 2015 Pierre Moulon.
+
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "openMVG/sfm/sfm_view.hpp"
+#include "third_party/stlplus3/filesystemSimplified/file_system.hpp"
+
+#include <cereal/types/polymorphic.hpp>
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_view.cpp]: alphabetical ordering

File: src/openMVG/sfm/sfm_view.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+namespace sfm {
+
+View::View(
+  const std::string & sImgPath,
+  IndexT view_id,
+  IndexT intrinsic_id,
+  IndexT pose_id,
+  IndexT width, 
+  IndexT height)
+  : s_Img_path(sImgPath)
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_view.cpp]: change to 
```
x,
y,
z
```
rather than
```
x
, y
```

File: src/openMVG/sfm/sfm_view.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+     cereal::make_nvp("width", ui_width),
+     cereal::make_nvp("height", ui_height),
+     cereal::make_nvp("id_view", id_view),
+     cereal::make_nvp("id_intrinsic", id_intrinsic),
+     cereal::make_nvp("id_pose", id_pose));
+
+  s_Img_path = stlplus::create_filespec(local_path, filename);
+}
+
+template void View::load
```
Comment (pmoulon) [inline src/openMVG/sfm/sfm_view.cpp]: alphabetical reordering based on cereal::X

File: src/software/VO/Monocular_VO.hpp [hunk 12]
```diff
@@ -11,6 +11,7 @@
 
 #include <deque>
 #include <set>
+#include <numeric>
```
Comment (pmoulon) [inline src/software/VO/Monocular_VO.hpp]: put numeric before set

**Comment by bjornpiltz** [comment]
Have a look at https://github.com/openMVG/openMVG/pull/911. I think that is a cleaner approach, achieving the same result.

**Comment by pmoulon** [comment]
Since #911 have been merged, I'm closing this for the moment. I think you will make another PR for the geometry module.

**Pull request closed without merging**